jplag/src/main/java/de/jplag/clustering/algorithm/GaussianProcess.java

```diff
@@ -0,0 +1,210 @@
+
+        return mu * this.standardDeviation + this.mean;
+    }
+
+    /**
+     * @param x position to predict at
+     * @return array containing the predicted [mean, std] at x.
+     */
+    public double[] predictWidthStd(RealVector x) {
+        RealVector k = maternKernel(X, x, lengthScale);
+        RealVector Kx = cholesky.getSolver().solve(k);
+
+        double mu = weight.dotProduct(k);
+        double sd = Math.sqrt(maternKernel(x, x, lengthScale) - Kx.dotProduct(k));
+
+        double[] out = new double[2];
+        out[0] = mu * this.standardDeviation + this.mean;
+        out[1] = sd * this.standardDeviation;
+
+        if (Double.isNaN(out[1])) {
+            out[1] = 0;
+        }
+
+        return out;
+    }
+
+    private static final double SQRT_5 = Math.sqrt(5);
+
+    /**
+     * Identical to {@link GaussianProcess#fit(List, double[], double, boolean, double[])} but using a one vector as length
+     * scale.
+     * @param X
+     * @param Y expected to have zero mean, unit variance if normalize is false
+     * @param noise variance of noise in Y
+     * @param normalize for Y
+     */
+    public static GaussianProcess fit(List<RealVector> X, double[] Y, double noise, boolean normalize) {
+        return fit(X, Y, noise, normalize, new ArrayRealVector(X.get(0).getDimension(), 1).toArray());
+    }
+
+    /**
+     * Fit Gaussian Process using a matern kernel.
+     * @param X
+     * @param Y expected to have zero mean, unit variance if normalize is false
+     * @param noise variance of noise in Y
+     * @param normalize if Y should be normalized
+     * @param lengthScale X values are divided by these values before calculating their euclidean distance. If all entries
+     * are equal the kernel is isometric.
+     */
+    public static GaussianProcess fit(List<RealVector> X, double[] Y, double noise, boolean normalize, double[] lengthScale) {
+        if (X.size() != Y.length) {
+            throw new IllegalArgumentException(MessageFormat.format("X and Y are of different dimensions {0} and {1}", X.size(), Y.length));
+        }
+        OptionalInt brokenXIndex = IntStream.range(1, X.size()).filter(i -> X.get(i).getDimension() != X.get(i - 1).getDimension()).findFirst();
+        brokenXIndex.ifPresent(brokenIndex -> {
+            throw new IllegalArgumentException(MessageFormat.format("X has different dimensions at index {0} ({2}) and index {1} ({3})",
+                    brokenIndex - 1, brokenXIndex, X.get(brokenIndex - 1).getDimension(), X.get(brokenIndex).getDimension()));
+        });
+        if (noise <= 0) {
+            throw new IllegalArgumentException(MessageFormat.format("noise must be strictly positive, got {0}", noise));
+        }
+        if (lengthScale.length != X.get(0).getDimension()) {
+            throw new IllegalArgumentException(
+                    MessageFormat.format("lengthScale is of different dimension {0} than X values {1}", lengthScale.length, X.get(0).getDimension()));
```

Comment:
Multiple Index Out of bounds exceptions and other exceptions might be possible if X or Y ist empty or null

Reply:
I've added a check for empty X at the beginning. And I mean in calling this with empty lists / null can not work anyway.